Add a functionality to produce canonical conformances and substitutions to fix a bug with SILBoxType’s generic arguments

swiftix opened the pull request:
Fix a bug with SILBoxType’s generic arguments

The set of SILBoxType’s generic arguments should be canonical, otherwise getCanonicalType() produces wrong results.

A canonical conformance is defined as a conformance which:
- does not contain any non-canonical types.
- Its type and interface type should be canonical.
- Any referenced conformances should be canonical.
- Any used substitutions should be canonical as well.

A substitution is canonical if:
- its replacement type is canonical
- all of its conformances are canonical

Resolves one of the issues in rdar://31603112

swiftix committed:
Add a functionality to produce canonical conformances and substitutions

A canonical conformance is defined as a conformance which:
- does not contain any non-canonical types.
- Its type and interface type should be canonical.
- Any referenced conformances should be canonical.
- Any used substitutions should be canonical as well.

A substitution is canonical if:
- its replacement type is canonical
- all of its conformances are canonical

swiftix committed:
Fix a bug with SILBoxType’s generic arguments

The set of SILBoxType’s generic arguments should be canonical, otherwise getCanonicalType() produces wrong results.

It does not contain a reduced test-case, because the actual bug happens when compiling the stdlib on one of the experimental branches and requires a rather complex setup. I'll try to provide a test-case later.

Resolves one of the issues in rdar://31603112

swiftix commented:
@swift-ci please smoke test

swiftix commented:
@DougGregor Doug, could you have a look?

slavapestov commented:
It sounds like a better approach would be to construct SILBoxType's substitutions from canonical types upfront. Canonicalizing conformances after the facts looks like it might be tricky to get right. Also @jckarter should take a look.

DougGregor commented:
We'll need the canonicalization of conformances anyway when we teach `BoundGenericType` to keep a proper `SubstitutionList` rather than it's array of `Type` arguments. The places that construct `SILBoxType` can pre-canonicalize (I guess), although it's not clear to me that `SILBoxType` has to always be canonical. Why not make it canonicalize like other types, @jckarter  ?

DougGregor commented on lib/AST/ProtocolConformance.cpp:
```diff
@@ -1059,3 +1059,151 @@ DeclContext::getLocalConformances(
+    auto normalConformance = cast<NormalProtocolConformance>(this);
+    auto sigConformances = normalConformance->getSignatureConformances();
+    for (auto conf : sigConformances)
+      if (!conf.isCanonical())
```
I think we want to define `NormalProtocolConformance` as always being canonical, because they are unique; it doesn't matter that we might have type sugar in the signature conformances.

DougGregor commented on lib/AST/ProtocolConformance.cpp:
```diff
@@ -1059,3 +1059,151 @@ DeclContext::getLocalConformances(
+    if (!genericConformance->isCanonical())
+      return false;
+    auto specSubs = spec->getGenericSubstitutions();
+    for (auto sub : specSubs) {
```
`const auto &sub`, please

DougGregor commented on lib/AST/ProtocolConformance.cpp:
```diff
@@ -1059,3 +1059,151 @@ DeclContext::getLocalConformances(
+      return false;
+    auto specSubs = spec->getGenericSubstitutions();
+    for (auto sub : specSubs) {
+      if (!sub.getReplacement()->isCanonical())
```
Maybe there should be a `Substitution::isCanonical()` instead of open coding it here?

swiftix committed:
Address review comments

slavapestov commented on include/swift/AST/Substitution.h:
```diff
@@ -48,6 +48,9 @@ class Substitution {
   
   Substitution(Type Replacement, ArrayRef<ProtocolConformanceRef> Conformance);
 
+  /// Checks whether the current substituion is canonical.
```
Typo: substituion

slavapestov commented on lib/AST/ProtocolConformance.cpp:
```diff
@@ -1059,3 +1059,151 @@ DeclContext::getLocalConformances(
+bool ProtocolConformance::isCanonical() const {
+  // Normal conformances are considered canonical by
+  // construction.
+  if (getKind() != ProtocolConformanceKind::Normal) {
```
If the kind is normal here, you can just return true.

slavapestov commented on lib/AST/ProtocolConformance.cpp:
```diff
@@ -1059,3 +1059,151 @@ DeclContext::getLocalConformances(
+    return true;
+  }
+  case ProtocolConformanceKind::Inherited: {
+    // Substitute the base.
```
The comments are wrong here

slavapestov commented on lib/AST/ProtocolConformance.cpp:
```diff
@@ -1059,3 +1059,151 @@ DeclContext::getLocalConformances(
+  bool subListWasCanonical = true;
+  for (auto &sub : subs) {
+    bool subWasCanonical = false;
+    auto canSub = sub.getCanonicalSubstitution(&subWasCanonical);
```
How about just checking isCanonical() on the substitution first and removing the out parameter. The conformance array will have zero or one elements most cases, so checking it is not a big deal.

slavapestov commented on lib/AST/ProtocolConformance.cpp:
```diff
@@ -1059,3 +1059,151 @@ DeclContext::getLocalConformances(
+
+  switch (getKind()) {
+  case ProtocolConformanceKind::Normal: {
+    //assert(isCanonical() && "Normal conformances should always be canonical");
```
Why not leave this in?

swiftix replied:
The NormalConformances are canonical by construction (see how NormalProtocolConformance::Profile is implemented. It uses only the protocol for hashing, everything else is derived from it), but the NormalConformance::getType and NormalConformance::getInterfaceType are not necessarily canonical, because they are derived and do no participate in profiling. Thus we cannot really  leave this check in.

slavapestov replied:
In that case isCanonical should just ignore the type and interface type of normal conformances.

Note that for specialized conformances, the type is also derivable from the substitutions and the type of the normal conformance. So either way it doesn't need to participate in profiling.

slavapestov commented on lib/AST/ProtocolConformance.cpp:
```diff
@@ -1059,3 +1059,151 @@ DeclContext::getLocalConformances(
+    auto spec = cast<SpecializedProtocolConformance>(this);
+    auto genericConformance = spec->getGenericConformance();
+    auto specSubs = spec->getGenericSubstitutions();
+    auto canSpecSubs = getCanonicalSubstitutionList(specSubs);
```
This allocates an extra copy of the list because getSpecializedConformance() copies it inside the conformance. Change getCanonicalSubstitutionList() to take a SmallVector mutable reference instead.

swiftix replied:
@slavapestov I see your point. But if I do that, then we always would need to copy the canonicalized conformances into this SmallVector. In the current version the assumption is that in most cases the conformances and list of conformances are canonicalized already and in such cases, no new allocations are performed at all. Only if there there is a need to canonicalize we do create a copy today. And we don't touch the existing substitution list, i.e. we do not change it in place.

But may be we should just allow the modification of the exiting substitution list in place? Then getCanonicalSubstitutionList may become canonicalSubstitutionList which would take a MutableArrayRef<Substitution> and would update the non-canonical elements by their canonicalized versions?

slavapestov reviewed (commented):
I'm worried about recursive conformances. @DougGregor can there ever be a cycle from a conformance to itself if you follow the substitutions?

slavapestov commented:
@swiftix Take a look at lib/SIL/TypeLowering.cpp, in getInterfaceBoxTypeForCapture(). This is where SILBoxTypes are constructed. The canonicalizeSubstitutions() method there can go away.

Also, look at lib/AST/Type.cpp, substType(). Here we canonicalize the replacement types but not the conformances. This might be the root cause of the bug you were hitting:

```
      for (auto &arg : newSubs) {
        arg = Substitution(arg.getReplacement()->getCanonicalType(),
                           arg.getConformances());
      }
```

slavapestov commented on lib/AST/ASTContext.cpp:
```diff
@@ -4187,8 +4187,12 @@ CanSILBoxType SILBoxType::get(ASTContext &C,
   llvm::FoldingSetNodeID id;
+
+  // Canonicalize substitutions.
+  Args = getCanonicalSubstitutionList(Args);
```
It looks like here we allocate memory for the substitutions inside the SILBoxType also, so an out parameter vector would be better than having getCanonicalSubstitutionList() allocate from context.

dabrahams commented:
@swiftix I am merging this into the unicode-rethink branch, so please make any amendments by adding commits to your branch rather than by rebasing.  Thanks!

slavapestov commented:
@dabrahams That's ridiculous. Why don't you rebase your branch if master diverges? This patch isn't fully baked yet.

dabrahams commented:
@slavapestov I am not working on this branch by myself, among many other reasons.

jckarter commented:
I thought that conformances were already fundamentally canonicalized at the AST level, since we use their pointer identity to identify the conformance across SIL.

jckarter commented:
In other words, only the replacement type of substitutions ought to need canonicalization, because the conformance should be canonical already.

swiftix commented:
@jckarter It turned out that conformances are not always canonicalized automatically. But if I understand you correctly, semantically it is fine to start with the canonicalized set of conformances, i.e. you don't need the syntactic sugar in the SILBoxTypes, right? In this case, we have multiple alternatives:
- I can just replace the existing canonicalizeSubstitutions call by a call of my version of the same thing that also canonicalizes the conformances.
- I can skip this step in getInterfaceBoxTypeForCapture and do it in SILBoxType::get

@DougGregor It could be even better to (conditionally, based on a flag) canonicalize conformances during the construction of the SubstitutionMaps to avoid a later step of walking the conformances and canonicalizing them. WDYT? Would it be possible? Or is it better to do it in a separate step like I do it in this patch?

swiftix committed:
Address review comments

- Avoid unnecessary memory allocation
- Fix spelling errors
- Some further minor fixes

swiftix commented:
@swift-ci please smoke test

swiftix commented:
@swift-ci please smoke test

swiftix commented:
@slavapestov @DougGregor I think I addressed all of your comments and suggestions.

slavapestov commented:
@jckarter Conformances cannot be canonicalized for correctness. Eg, `Array<((Int, Int)) -> ()>` and `Array<(Int, Int) -> ()>` must have two different conformances to `Sequence` so that we get the correct behavior when we call elements with arguments vs a tuple argument.

slavapestov commented:
@swiftix Looks good to me, thanks!

@dabrahams Please merge the updated patch into your branch also, it is mildly more correct.
